tests/component/test_time.py

```diff
@@ -0,0 +1,277 @@
+        (JAN_01_2022_HIGHTIME, 999999, 0xFFFFEF39085F4800),
+    ],
+)
+def test___libtimestamp___convert_microseconds_to_and_from_subseconds___succeeds(
+    from_dt, microsecond, subseconds
+):
+    precise_dt = from_dt.replace(microsecond=microsecond)
+    ts = LibTimestamp.from_datetime(precise_dt)
+
+    # whole seconds shouldn't change
+    assert ts.msb == JAN_01_2022_LIB.msb
+    assert ts.lsb == subseconds
+
+    # now convert back
+    to_dt = ts.to_datetime(tzinfo=timezone.utc)
+
+    to_dt_microsecond = to_dt.microsecond
+    # comparison is tricky since imprecision in the conversion to NI-BTF are
+    # caught by the higher precision values in hightime, so we round here.
+    if to_dt.femtosecond > LibTimestamp.MAX_FS / 2:
+        to_dt_microsecond += 1
+
+    assert to_dt_microsecond == microsecond
+
+
+@pytest.mark.parametrize(
+    "from_dt, femtosecond, subseconds",
+    [
+        (JAN_01_2022_HIGHTIME, 0, 0),
+        (JAN_01_2022_HIGHTIME, 1, 0x480F),
+    ],
+)
+def test___libtimestamp___convert_femtoseconds_to_and_from_subseconds___succeeds(
+    from_dt, femtosecond, subseconds
+):
+    precise_dt = from_dt.replace(femtosecond=femtosecond)
+    ts = LibTimestamp.from_datetime(precise_dt)
+
+    # whole seconds shouldn't change
+    assert ts.msb == JAN_01_2022_LIB.msb
+    assert ts.lsb == subseconds
+
+    # now convert back
+    to_dt = ts.to_datetime(tzinfo=timezone.utc)
+    to_dt_femtosecond = to_dt.femtosecond
+    # comparison is tricky since imprecision in the conversion to NI-BTF are
+    # caught by the higher precision values in hightime, so we round here.
+    if to_dt.yoctosecond > LibTimestamp.MAX_YS / 2:
+        to_dt_femtosecond += 1
+    assert to_dt_femtosecond == femtosecond
+
+
+@pytest.mark.parametrize(
+    "from_dt, yoctosecond, subseconds, yoctosecond_round_trip",
+    [
+        (JAN_01_2022_HIGHTIME, 0, 0, 0),
+        # Yoctoseconds is quite a bit more precise than NI-BTF
+        (JAN_01_2022_HIGHTIME, 54210, 1, 54210),
+        (JAN_01_2022_HIGHTIME, 54211, 1, 54210),
+    ],
+)
+def test___libtimestamp___convert_yoctoseconds_to_and_from_subseconds___succeeds(
+    from_dt, yoctosecond, subseconds, yoctosecond_round_trip
+):
+    precise_dt = from_dt.replace(yoctosecond=yoctosecond)
+    ts = LibTimestamp.from_datetime(precise_dt)
+
+    # whole seconds shouldn't change
+    assert ts.msb == JAN_01_2022_LIB.msb
+    assert ts.lsb == subseconds
+
+    # now convert back
+    to_dt = ts.to_datetime(tzinfo=timezone.utc)
+    assert to_dt.yoctosecond == yoctosecond_round_trip
+
+
+# Note: I can't actually test overflow because Python's datetime object is
+# limited to years 1-9999, but the NI-BTF format can represent time before the
+# Big Bang and until about year 292 billion. Oh well.
+
+
+@pytest.mark.parametrize("from_dt", [(JAN_01_2022_DATETIME), (JAN_01_2022_HIGHTIME)])
+def test___grpctimestamp___convert_from_utc___succeeds(from_dt):
```

Comment:
The gRPC tests need to be skipped when gRPC isn't available

Reply:
They're skipped in the CI/PR since it runs with `not grpc`. Is there something else I should do explicitly?

Reply:
Oh, I see. I forgot the "base" tox configuration filters out grpc tests to avoid unnecessary skipped tests.

`@pytest.mark.grpc_only(...)` or `@pytest.mark.grpc_skip(...)` won't work because those markers are only for init_kwargs.

It sounds like `-k not grpc` matches against both filenames and test names, then I guess it's fine to leave it as-is.

If it only matched test names, you would still need the try/except around the gRPC imports so that pytest could import the file and find out the test names. BTW, the try/except could probably be used to skip the whole module:
```
try:
    import grpcio
except ImportError:
    pytest.skip(allow_module_level=True, reason="Requires gRPC")
```